### ISSUE TITLE
[Web Extension Transformer] doesn't support code splitting

### ISSUE THREAD

**Issue by fregante** [parcel-bundler/parcel#5859]
# 🙋 feature request

Code splitting via `import()` creates a `script` tag and injects it ([source](https://github.com/parcel-bundler/parcel/blob/7717e2547f510c89ea57bbfec62147bfb912ba49/packages/runtimes/js/src/loaders/browser/js-loader.js)) but this doesn't work for "content scripts" in extensions because such scripts are not run in the extension’s isolated world. Injecting `script` is actually a way to explicitly [escape the isolated world](https://stackoverflow.com/a/9517879/288906)

## 😯 Current Behavior

The code-splitted file doesn't have access to the `chrome.*` APIs like the rest of the `entry.js` file does.

## 💁 Possible Solution

I'm not aware of a way to load script from a content script in the same world (only non-content scripts pages can use the `chrome.tabs.executeScript` API to do that) so I think the only solution would be to:

1. Add the secondary file via manifest.json, before the entry, e.g.
	```diff
	  "content_scripts": [{
	  	"scripts": [
	+ 		"vendor.d186d19.js",
	  		"entry.js"
	  	]
	  }]
	```
2. Have parcel’s loader execute its code when `import()` is called.

## 🔦 Context

I'm considering content-splitting to avoid bundling the same dependency over and over. [In this PR](https://github.com/fregante/browser-extension-template/pull/44) the files are all 30KB+ because they all bundle their own version of a library:

```
distribution/background.js                                                     33.61 KB    199ms
distribution/options.a7e143c4.js                                               34.95 KB    689ms
```

but with `import()` I'd get this:

```
distribution/background.js                                                      8.57 KB    189ms
distribution/webext-options-sync.a0fe799f.js                                   35.79 KB    812ms
distribution/options.312654a3.js                                                9.92 KB    811ms
```

**Comment by fregante** [comment]
Because of the sync nature of the suggested solution, I wish there was a way to require code-splitted files synchronously (a-la-`require`), but I suppose that's not going to happen.

**Comment by mischnic** [comment]
Well there are two ways to achieve codesplitting:
- import() to create async child bundles
- shared bundles

The problem with the latter is https://github.com/parcel-bundler/parcel/issues/4227 (for the case of simply running `parcel build index.js`), but in this case there is already a manifest, so it might be possible to implement this.

**Comment by fregante** [comment]
> import() to create async child bundles

Yeah that’d be ideal, but:

- it’s not supported by Firefox in content scripts: https://bugzilla.mozilla.org/show_bug.cgi?id=1451545
- it’s not supported by either browsers as a background unless the user loads a background page instead (uncommon and otherwise unnecessary). Mentioned here https://stackoverflow.com/a/53033388/288906

> shared bundles

Yes 🙌, I think that’s what I suggested above.

**Comment by mischnic** [comment]
The main thing the prevents shared bundles is `isEntry: true`. Without that, there should be shared bundles. A webextension packager would be needed to then prepend the shared bundles into the `content_scripts` array (similar to how the HTML packager inserts `<script>` tags for shared bundles).

https://github.com/parcel-bundler/parcel/blob/f28eaf60ac018d49d2176c563c86d09f84feaa8e/packages/transformers/webextension/src/WebExtensionTransformer.js#L87

But I think there was some problem if it's set to false.

**Comment by 101arrowz** [comment]
The problem was the removal of the prelude from all descendants of the manifest when another descendant had received the prelude. Normally this is fine because they all share the same global environment, but in this case that doesn't happen. If that can be circumvented, `isEntry: true` is not necessary.

**Comment by mischnic** [comment]
>  A webextension packager would be needed to then prepend the shared bundles into the content_scripts array (similar to how the HTML packager inserts <script> tags for shared bundles).

 (That would also be needed for https://github.com/parcel-bundler/parcel/issues/5865)

**Comment by fregante** [comment]
Edit: maybe it still make sense as a code sharing mechanism

> A webextension packager would be needed to then prepend the shared bundles into the `content_scripts` array

That defeats the purpose of splitting it in the first place (other than just "supporting" `import()` in Parcel). You can see how this webpack plugin does this: https://github.com/crimx/webpack-target-webextension

I think:

- when a content script tries to load another file, it sends a message to the background script, and the background script uses `chrome.tabs.executeCode` to inject the new bundle
- when any other `chrome://` page tries to load another file, it does so via `script` tag and onload event

**Comment by fregante** [comment]
I have news that would make this much easier to implement: almost-raw `import()` calls are now supported in Firefox 89 and they work in every browser including Safari, in content scripts and in background pages 🎉 

https://bugzilla.mozilla.org/show_bug.cgi?id=1536094

```js
import(chrome.runtime.getURL('file.js')).then(mod => {console.log(mod.default)})
```
```json5
// manifest.json
{
  "web_accessible_resources": ["file.js"]
}
```

This works in Parcel **today** because the dynamic `import` call is ignored while `web_accessible_resources` pushes `file.js` into Parcel’s graph.

---

What's next for full code-spltting support:

- when encountering `import('left-pad')` calls, code-split as usual  BUT preserve the actual `import()` call  instead of using its own loading method; this is useful to every Parcel user (is there already an issue for this?)
- automatically wrap `import`’s path in `chrome.runtime.getURL()` + add the resulting bundle to the manifest


#### Example input

```js
// entry-a.js
import('left-pad').then(alert)
```

```js
// entry-b.js
import('left-pad').then(console.log)
```

#### Desired output

```js
// entry-a.js
import(chrome.runtime.getURL('left-pad.8f8a6d.js')).then(alert)
```

```js
// entry-b.js
import(chrome.runtime.getURL('left-pad.8f8a6d.js')).then(console.log)
```

```js
// left-pad.8f8a6d.js
export default (str, n) => str.padStart(n)
```

```json5
// manifest.json
{
  "web_accessible_resources": ["left-pad.8f8a6d.js"]
}
```

**Comment by 101arrowz** [comment]
I think #7050 unintentionally fixed this but someone should verify.

**Comment by 101arrowz** [comment]
Yep this should be fixed in the latest versions but if anyone uses dynamic import more extensively, please test it out and report back here.

**Comment by SpaceK33z** [comment]
> ```js
> import(chrome.runtime.getURL('file.js')).then(mod => {console.log(mod.default)})
> ```

I'm trying to do what @fregante is doing, but it's failing because Parcel always seems to rewrite `import()` to `require()`. Is there any way to tell Parcel to not rewrite import to require?

**Comment by fregante** [comment]
If that's the case, it means this was not fixed, as my last comment specifically suggested that `import()` should not be converted to `require()`. I haven't worked on Parcel extensions much since this PR was closed so I did not check. 

Can we reopen this issue? @101arrowz

**Comment by Curve** [comment]
> > ```js
> > import(chrome.runtime.getURL('file.js')).then(mod => {console.log(mod.default)})
> > ```
> 
> I'm trying to do what @fregante is doing, but it's failing because Parcel always seems to rewrite `import()` to `require()`. Is there any way to tell Parcel to not rewrite import to require?

I'm experiencing the exact same issue, is there some way to make parcel not rewrite the import? Something like webpacks `/*webpackIgnore: true*/` would be useful here.

**Comment by fregante** [comment]
It's probably best to create a new issue with updated repro.
